File: legged_gym/scripts/analyze_load_experiment.py
#!/usr/bin/env python3
"""Analyze the WF_TRON1A payload load-estimation experiment.

The CSV is expected to live beside this script. The script produces the same
three PNG outputs described by the original MATLAB README.
"""

from pathlib import Path
from typing import List

import matplotlib

matplotlib.use("Agg")

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd


SCRIPT_DIR = Path(__file__).resolve().parent
CSV_FILE = SCRIPT_DIR / "wf_deploy_log_real_load_fall_latest.csv"

LOAD_MASSES = np.array([4.425, 3.48, 4.50])
ADD_TIMES = np.array([135.4, 149.0, 158.1])
REMOVE_TIMES = np.array([166.0, 169.8, 177.6])
RIGHT_ADD_TIMES = np.array([182.5, 192.1])
FALL_TIME = 198.75
MASS_PLOT_XLIM = (130.0, 180.0)
MASS_EST_OFFSET = 1.0
BODY_MASS = 9.58
BODY_COM0 = np.array([0.04576, 0.00014, -0.16398])


def require_columns(table: pd.DataFrame, columns: List[str]) -> None:
    missing = [column for column in columns if column not in table.columns]
    if missing:
        raise ValueError(f"Missing required CSV columns: {', '.join(missing)}")


def add_vertical_marker(ax: plt.Axes, x: float, label: str, color: str, style: str,
                        y: float = 0.98, va: str = "top") -> None:
    ax.axvline(x, color=color, linestyle=style, linewidth=1.0)
    offset_y = -4 if va == "top" else 4
    ax.annotate(
        label,
        xy=(x, y),
        xycoords=("data", "axes fraction"),
        xytext=(3, offset_y),
        textcoords="offset points",
        ha="left",
        va=va,
        fontsize=8,
        color=color,
    )


def marker_is_visible(x: float, xlim: tuple) -> bool:
    return xlim[0] <= x <= xlim[1]


def add_load_markers(ax: plt.Axes, include_right_adds: bool = False,
                     xlim: tuple = None) -> None:
    for index, add_time in enumerate(ADD_TIMES, start=1):
        if xlim is not None and not marker_is_visible(add_time, xlim):
            continue
        add_vertical_marker(ax, add_time, f"add {index}", "black", "--")
    for index, (remove_index, remove_time) in enumerate(zip([3, 2, 1], REMOVE_TIMES)):
        if xlim is not None and not marker_is_visible(remove_time, xlim):
            continue
        add_vertical_marker(
            ax,
            remove_time,
            f"remove {remove_index}",
            "m",
            "--",
            y=0.98 - 0.07 * (index % 2),
        )
    if include_right_adds:
        for index, add_time in enumerate(RIGHT_ADD_TIMES, start=1):
            if xlim is not None and not marker_is_visible(add_time, xlim):
                continue
            add_vertical_marker(
                ax,
                add_time,
                f"right add {index}",
                "black",
                ":",
                y=0.08 + 0.08 * (index - 1),
                va="bottom",
            )
    if xlim is None or marker_is_visible(FALL_TIME, xlim):
        add_vertical_marker(ax, FALL_TIME, "fall", "r", "--")


def save_figure(fig: plt.Figure, filename: str) -> Path:
    output_path = SCRIPT_DIR / filename
    fig.tight_layout()
    fig.savefig(output_path, dpi=200)
    plt.close(fig)
    return output_path


def time_weighted_mean(t: np.ndarray, values: np.ndarray, start: float, end: float) -> float:
    idx = (t > start) & (t < end)
    t_window = np.concatenate(([start], t[idx], [end]))
    if t_window.size < 2:
        return float("nan")
    values_window = np.interp(t_window, t, values)
    return float(np.trapz(values_window, t_window) / (end - start))


def actual_loaded_mean() -> float:
    phase_starts = np.array([ADD_TIMES[0], ADD_TIMES[1], ADD_TIMES[2],
                             REMOVE_TIMES[0], REMOVE_TIMES[1]])
    phase_ends = np.array([ADD_TIMES[1], ADD_TIMES[2], REMOVE_TIMES[0],
                           REMOVE_TIMES[1], REMOVE_TIMES[2]])
    phase_masses = np.array([
        LOAD_MASSES[0],
        LOAD_MASSES[0] + LOAD_MASSES[1],
        LOAD_MASSES.sum(),
        LOAD_MASSES[0] + LOAD_MASSES[1],
        LOAD_MASSES[0],
    ])
    return float(np.sum((phase_ends - phase_starts) * phase_masses)
                 / (REMOVE_TIMES[-1] - ADD_TIMES[0]))


def infer_payload_position_from_robot_com_delta(
    com_delta: np.ndarray,
    payload_mass: np.ndarray,
) -> np.ndarray:
    payload_pos = np.full_like(com_delta, np.nan, dtype=float)
    valid = payload_mass > 1e-6
    scale = (BODY_MASS + payload_mass[valid]) / payload_mass[valid]
    payload_pos[valid] = BODY_COM0 + scale[:, None] * com_delta[valid]
    return payload_pos


def main() -> None:
    if not CSV_FILE.exists():
        raise FileNotFoundError(f"CSV file not found: {CSV_FILE}")

    table = pd.read_csv(CSV_FILE)
    require_columns(table, ["t", "mass_est", "comx", "comy", "comz"])

    t = table["t"].to_numpy(dtype=float)
    mass_est_raw = table["mass_est"].to_numpy(dtype=float)
    mass_est = mass_est_raw + MASS_EST_OFFSET
    comx = table["comx"].to_numpy(dtype=float)
    comy = table["comy"].to_numpy(dtype=float)
    comz = table["comz"].to_numpy(dtype=float)
    robot_com_delta = np.column_stack((comx, comy, comz))

    mass_smooth = (
        pd.Series(mass_est)
        .rolling(window=51, center=True, min_periods=1)
        .median()
        .to_numpy()
    )

    actual_times = np.array([t.min(), *ADD_TIMES, *REMOVE_TIMES, t.max()])
    actual_mass = np.array([
        0.0,
        LOAD_MASSES[0],
        LOAD_MASSES[0] + LOAD_MASSES[1],
        LOAD_MASSES.sum(),
        LOAD_MASSES[0] + LOAD_MASSES[1],
        LOAD_MASSES[0],
        0.0,
        0.0,
    ])

    fig, ax = plt.subplots(figsize=(12, 6))
    ax.plot(t, mass_est, color=(0.75, 0.75, 0.75),
            label=f"estimated raw +{MASS_EST_OFFSET:.1f} kg")
    ax.plot(t, mass_smooth, color="b", linewidth=1.5,
            label=f"estimated movmedian +{MASS_EST_OFFSET:.1f} kg")
    ax.step(actual_times, actual_mass, where="post", color="r", linewidth=2.0,
            label="actual payload")
    add_load_markers(ax, xlim=MASS_PLOT_XLIM)
    ax.grid(True)
    ax.set_xlabel("time [s]")
    ax.set_ylabel("mass [kg]")
    ax.set_title("Estimated payload mass vs actual payload mass")
    ax.set_xlim(MASS_PLOT_XLIM)
    ax.legend(loc="lower center")
    mass_png = save_figure(fig, "payload_mass_comparison.png")

    print("\nApproximate middle-load mass comparison")
    print(
        "Assumed load masses: "
        f"{LOAD_MASSES[0]:.3f}, {LOAD_MASSES[1]:.3f}, {LOAD_MASSES[2]:.3f} kg"
    )
    print(
        "Add times: "
        f"{ADD_TIMES[0]:.1f}, {ADD_TIMES[1]:.1f}, {ADD_TIMES[2]:.1f} s; "
        "remove times, last-in first-out: "
        f"{REMOVE_TIMES[0]:.1f}, {REMOVE_TIMES[1]:.1f}, {REMOVE_TIMES[2]:.1f} s\n"
    )
    print(f"Applied estimated mass offset: +{MASS_EST_OFFSET:.3f} kg\n")
    print(f"{'window':>11s} {'actual_kg':>14s} {'est_med_kg':>14s} {'error_kg':>14s}")

    plateau_starts = ADD_TIMES
    plateau_ends = np.array([ADD_TIMES[1], ADD_TIMES[2], REMOVE_TIMES[0]])
    actual_plateaus = np.cumsum(LOAD_MASSES)

    for start, end, actual in zip(plateau_starts, plateau_ends, actual_plateaus):
        t0 = start + 2.0
        t1 = min(end - 1.0, t0 + 8.0)
        idx = (t >= t0) & (t <= t1)
        est_med = np.nanmedian(mass_est[idx])
        print(f"{t0:5.1f}-{t1:<5.1f} {actual:14.3f} {est_med:14.3f} {est_med - actual:14.3f}")

    loaded_start = ADD_TIMES[0]
    loaded_end = REMOVE_TIMES[-1]
    actual_mean = actual_loaded_mean()
    raw_mean = time_weighted_mean(t, mass_est, loaded_start, loaded_end)
    smooth_mean = time_weighted_mean(t, mass_smooth, loaded_start, loaded_end)
    print("\nWhole loaded-period mean bias")
    print(f"Loaded period: {loaded_start:.1f}-{loaded_end:.1f} s")
    print(f"Actual red mean:          {actual_mean:8.3f} kg")
    print(f"Estimated raw mean:       {raw_mean:8.3f} kg, residual {actual_mean - raw_mean:8.3f} kg")
    print(
        f"Estimated movmedian mean: {smooth_mean:8.3f} kg, "
        f"residual {actual_mean - smooth_mean:8.3f} kg"
    )

    payload_pos = infer_payload_position_from_robot_com_delta(robot_com_delta, mass_est)
    loaded_mask = (t >= ADD_TIMES[0]) & (t < REMOVE_TIMES[-1])
    payload_pos_plot = payload_pos.copy()
    payload_pos_plot[~loaded_mask] = np.nan

    print("\nPayload position inferred from robot COM delta")
    print(
        "Formula: payload_pos = body_com0 + "
        "(body_mass + payload_mass) / payload_mass * robot_com_delta"
    )
    print(f"body_mass = {BODY_MASS:.3f} kg, body_com0 = {BODY_COM0.tolist()}")
    print(f"{'window':>11s} {'load_x_m':>12s} {'load_y_m':>12s} {'load_z_m':>12s}")
    for start, end in zip(plateau_starts, plateau_ends):
        t0 = start + 2.0
        t1 = min(end - 1.0, t0 + 8.0)
        idx = (t >= t0) & (t <= t1)
        med = np.nanmedian(payload_pos[idx], axis=0)
        print(f"{t0:5.1f}-{t1:<5.1f} {med[0]:12.3f} {med[1]:12.3f} {med[2]:12.3f}")

    fig, ax = plt.subplots(figsize=(12, 6))
    ax.plot(t, comx, linewidth=1.2, label="comx")
    ax.plot(t, comy, linewidth=1.2, label="comy")
    ax.plot(t, comz, linewidth=1.2, label="comz")
    add_load_markers(ax, include_right_adds=True)
    ax.grid(True)
    ax.set_xlabel("time [s]")
    ax.set_ylabel("estimated robot COM offset [m]")
    ax.set_title("Estimated robot COM offset coordinates")
    ax.legend(loc="best")
    com_png = save_figure(fig, "payload_com_coordinates.png")

    fig, ax = plt.subplots(figsize=(7, 7))
    scatter = ax.scatter(comx, comy, c=t, s=8, cmap="viridis")
    ax.plot(comx, comy, color=(0.7, 0.7, 0.7, 0.35))
    ax.plot(comx[0], comy[0], "go", markersize=8, linewidth=1.5, label="start")
    fall_idx = int(np.argmin(np.abs(t - FALL_TIME)))
    ax.plot(comx[fall_idx], comy[fall_idx], "rx", markersize=10, linewidth=2.0,
            label="near fall")
    ax.grid(True)
    ax.axis("equal")
    fig.colorbar(scatter, ax=ax, label="time [s]")
    ax.set_xlabel("robot COM delta x [m]")
    ax.set_ylabel("robot COM delta y [m]")
    ax.set_title("Estimated robot COM offset x-y trajectory, color = time [s]")
    ax.legend(loc="best")
    xy_png = save_figure(fig, "payload_com_xy_trajectory.png")

    fig, ax = plt.subplots(figsize=(12, 6))
    ax.plot(t, payload_pos_plot[:, 0], linewidth=1.2, label="payload x inferred")
    ax.plot(t, payload_pos_plot[:, 1], linewidth=1.2, label="payload y inferred")
    ax.plot(t, payload_pos_plot[:, 2], linewidth=1.2, label="payload z inferred")
    add_load_markers(ax, xlim=MASS_PLOT_XLIM)
    ax.grid(True)
    ax.set_xlabel("time [s]")
    ax.set_ylabel("estimated payload position [m]")
    ax.set_title("Payload position inferred from robot COM offset")
    ax.set_xlim(MASS_PLOT_XLIM)
    ax.legend(loc="best")
    payload_pos_png = save_figure(fig, "payload_position_from_robot_com.png")

    print("\nSaved figures:")
    for output_path in [mass_png, com_png, xy_png, payload_pos_png]:
        print(f"  {output_path}")


if __name__ == "__main__":
    main()

File: legged_gym/scripts/analyze_load_experiment.m
% Analyze WF_TRON1A load estimation experiment.
% Put this script in the same folder as:
%   wf_deploy_log_real_load_fall_latest.csv
%
% Edit the event times below if you want to align the actual load steps
% more precisely with your video or notes.

clear; clc; close all;

scriptDir = fileparts(mfilename('fullpath'));
if isempty(scriptDir)
    scriptDir = pwd;
end

csvFile = fullfile(scriptDir, 'wf_deploy_log_real_load_fall_latest.csv');
T = readtable(csvFile);

t = T.t;
massEst = T.mass_est;
comx = T.comx;
comy = T.comy;
comz = T.comz;

% Actual payloads for the middle "add three loads" test.
% Assumption: "4,425 kg" means 4.425 kg. Change to 4.25 if needed.
loadMasses = [4.425, 3.48, 4.50];

% Approximate event times inferred from jumps in massEst.
% t = 135.4: first load appears
% t = 149.0: second load appears
% t = 158.1: third load appears
% t = 166.0, 169.8, 177.6: loads are removed in last-in first-out order
addTimes = [135.4, 149.0, 158.1];
removeTimes = [166.0, 169.8, 177.6];

% Later right-side load test markers, inferred from the same log.
rightAddTimes = [182.5, 192.1];
fallTime = 198.75;
massPlotXLim = [130.0, 180.0];
massEstOffset = 1.0;
bodyMass = 9.58;
bodyCom0 = [0.04576, 0.00014, -0.16398];

massEst = massEst + massEstOffset;

actualTimes = [min(t), addTimes, removeTimes, max(t)];
actualMass = [ ...
    0, ...
    loadMasses(1), ...
    loadMasses(1) + loadMasses(2), ...
    sum(loadMasses), ...
    loadMasses(1) + loadMasses(2), ...
    loadMasses(1), ...
    0, ...
    0 ...
];

massSmooth = movmedian(massEst, 51, 'omitnan');

%% Figure 1: estimated mass vs actual payload mass
figure('Name', 'Payload mass comparison', 'Color', 'w');
plot(t, massEst, 'Color', [0.75 0.75 0.75], ...
    'DisplayName', sprintf('estimated raw +%.1f kg', massEstOffset));
hold on;
plot(t, massSmooth, 'b', 'LineWidth', 1.5, ...
    'DisplayName', sprintf('estimated movmedian +%.1f kg', massEstOffset));
stairs(actualTimes, actualMass, 'r', 'LineWidth', 2.0, 'DisplayName', 'actual payload');

for k = 1:numel(addTimes)
    if addTimes(k) >= massPlotXLim(1) && addTimes(k) <= massPlotXLim(2)
        xline(addTimes(k), '--k', sprintf('add %d', k), ...
            'LabelOrientation', 'horizontal', 'LabelVerticalAlignment', 'bottom');
    end
end
removeLabels = [3, 2, 1];
for k = 1:numel(removeTimes)
    if removeTimes(k) >= massPlotXLim(1) && removeTimes(k) <= massPlotXLim(2)
        xline(removeTimes(k), '--m', sprintf('remove %d', removeLabels(k)), ...
            'LabelOrientation', 'horizontal', 'LabelVerticalAlignment', 'bottom');
    end
end
if fallTime >= massPlotXLim(1) && fallTime <= massPlotXLim(2)
    xline(fallTime, '--r', 'fall', ...
        'LabelOrientation', 'horizontal', 'LabelVerticalAlignment', 'bottom');
end

grid on;
xlabel('time [s]');
ylabel('mass [kg]');
title('Estimated payload mass vs actual payload mass');
xlim(massPlotXLim);
legend('Location', 'south');

%% Print rough plateau errors.
fprintf('\nApproximate middle-load mass comparison\n');
fprintf('Assumed load masses: %.3f, %.3f, %.3f kg\n', loadMasses);
fprintf(['Add times: %.1f, %.1f, %.1f s; remove times, ', ...
    'last-in first-out: %.1f, %.1f, %.1f s\n\n'], ...
    addTimes, removeTimes);
fprintf('Applied estimated mass offset: +%.3f kg\n\n', massEstOffset);
fprintf('%8s %14s %14s %14s\n', 'window', 'actual_kg', 'est_med_kg', 'error_kg');

plateauStarts = [addTimes(1), addTimes(2), addTimes(3)];
plateauEnds = [addTimes(2), addTimes(3), removeTimes(1)];
actualPlateaus = cumsum(loadMasses);

for k = 1:numel(plateauStarts)
    % Use the middle of each plateau, avoiding the first two seconds of hand motion.
    t0 = plateauStarts(k) + 2.0;
    t1 = min(plateauEnds(k) - 1.0, t0 + 8.0);
    idx = t >= t0 & t <= t1;
    estMed = median(massEst(idx), 'omitnan');
    fprintf('%5.1f-%5.1f %14.3f %14.3f %14.3f\n', ...
        t0, t1, actualPlateaus(k), estMed, estMed - actualPlateaus(k));
end

%% Print whole loaded-period mean bias.
loadedStart = addTimes(1);
loadedEnd = removeTimes(end);
phaseStarts = [addTimes(1), addTimes(2), addTimes(3), removeTimes(1), removeTimes(2)];
phaseEnds = [addTimes(2), addTimes(3), removeTimes(1), removeTimes(2), removeTimes(3)];
phaseMasses = [ ...
    loadMasses(1), ...
    loadMasses(1) + loadMasses(2), ...
    sum(loadMasses), ...
    loadMasses(1) + loadMasses(2), ...
    loadMasses(1) ...
];
actualMean = sum((phaseEnds - phaseStarts) .* phaseMasses) / (loadedEnd - loadedStart);

meanTimes = [loadedStart; t(t > loadedStart & t < loadedEnd); loadedEnd];
massMeanValues = interp1(t, massEst, meanTimes, 'linear');
smoothMeanValues = interp1(t, massSmooth, meanTimes, 'linear');
rawMean = trapz(meanTimes, massMeanValues) / (loadedEnd - loadedStart);
smoothMean = trapz(meanTimes, smoothMeanValues) / (loadedEnd - loadedStart);

fprintf('\nWhole loaded-period mean bias\n');
fprintf('Loaded period: %.1f-%.1f s\n', loadedStart, loadedEnd);
fprintf('Actual red mean:          %8.3f kg\n', actualMean);
fprintf('Estimated raw mean:       %8.3f kg, residual %8.3f kg\n', ...
    rawMean, actualMean - rawMean);
fprintf('Estimated movmedian mean: %8.3f kg, residual %8.3f kg\n', ...
    smoothMean, actualMean - smoothMean);

%% Infer payload position from robot COM delta.
robotComDelta = [comx, comy, comz];
payloadPos = nan(size(robotComDelta));
validPayloadMass = massEst > 1e-6;
scale = (bodyMass + massEst(validPayloadMass)) ./ massEst(validPayloadMass);
payloadPos(validPayloadMass, :) = repmat(bodyCom0, sum(validPayloadMass), 1) ...
    + robotComDelta(validPayloadMass, :) .* repmat(scale, 1, 3);
loadedMask = t >= addTimes(1) & t < removeTimes(end);
payloadPosPlot = payloadPos;
payloadPosPlot(~loadedMask, :) = nan;

fprintf('\nPayload position inferred from robot COM delta\n');
fprintf(['Formula: payload_pos = body_com0 + ', ...
    '(body_mass + payload_mass) / payload_mass * robot_com_delta\n']);
fprintf('body_mass = %.3f kg, body_com0 = [%.5f %.5f %.5f]\n', ...
    bodyMass, bodyCom0);
fprintf('%8s %12s %12s %12s\n', 'window', 'load_x_m', 'load_y_m', 'load_z_m');
for k = 1:numel(plateauStarts)
    t0 = plateauStarts(k) + 2.0;
    t1 = min(plateauEnds(k) - 1.0, t0 + 8.0);
    idx = t >= t0 & t <= t1;
    med = median(payloadPos(idx, :), 1, 'omitnan');
    fprintf('%5.1f-%5.1f %12.3f %12.3f %12.3f\n', ...
        t0, t1, med(1), med(2), med(3));
end

%% Figure 2: estimated COM coordinates over time
figure('Name', 'Estimated COM coordinates', 'Color', 'w');
plot(t, comx, 'LineWidth', 1.2, 'DisplayName', 'comx');
hold on;
plot(t, comy, 'LineWidth', 1.2, 'DisplayName', 'comy');
plot(t, comz, 'LineWidth', 1.2, 'DisplayName', 'comz');

for k = 1:numel(addTimes)
    xline(addTimes(k), '--k', sprintf('add %d', k), ...
        'LabelOrientation', 'horizontal', 'LabelVerticalAlignment', 'bottom');
end
for k = 1:numel(removeTimes)
    xline(removeTimes(k), '--m', sprintf('remove %d', removeLabels(k)), ...
        'LabelOrientation', 'horizontal', 'LabelVerticalAlignment', 'bottom');
end
for k = 1:numel(rightAddTimes)
    xline(rightAddTimes(k), ':k', sprintf('right add %d', k), ...
        'LabelOrientation', 'horizontal', 'LabelVerticalAlignment', 'top');
end
xline(fallTime, '--r', 'fall', ...
    'LabelOrientation', 'horizontal', 'LabelVerticalAlignment', 'bottom');

grid on;
xlabel('time [s]');
ylabel('estimated robot COM offset [m]');
title('Estimated robot COM offset coordinates');
legend('Location', 'best');

%% Figure 3: COM x-y trajectory
figure('Name', 'Estimated COM xy trajectory', 'Color', 'w');
scatter(comx, comy, 8, t, 'filled');
hold on;
plot(comx, comy, 'Color', [0.7 0.7 0.7 0.35]);
plot(comx(1), comy(1), 'go', 'MarkerSize', 8, 'LineWidth', 1.5, ...
    'DisplayName', 'start');
[~, fallIdx] = min(abs(t - fallTime));
plot(comx(fallIdx), comy(fallIdx), 'rx', 'MarkerSize', 10, 'LineWidth', 2.0, ...
    'DisplayName', 'near fall');
grid on;
axis equal;
colorbar;
xlabel('robot COM delta x [m]');
ylabel('robot COM delta y [m]');
title('Estimated robot COM offset x-y trajectory, color = time [s]');
legend('Location', 'best');

%% Figure 4: payload position inferred from robot COM delta
figure('Name', 'Payload position inferred from robot COM', 'Color', 'w');
plot(t, payloadPosPlot(:, 1), 'LineWidth', 1.2, 'DisplayName', 'payload x inferred');
hold on;
plot(t, payloadPosPlot(:, 2), 'LineWidth', 1.2, 'DisplayName', 'payload y inferred');
plot(t, payloadPosPlot(:, 3), 'LineWidth', 1.2, 'DisplayName', 'payload z inferred');

for k = 1:numel(addTimes)
    if addTimes(k) >= massPlotXLim(1) && addTimes(k) <= massPlotXLim(2)
        xline(addTimes(k), '--k', sprintf('add %d', k), ...
            'LabelOrientation', 'horizontal', 'LabelVerticalAlignment', 'bottom');
    end
end
for k = 1:numel(removeTimes)
    if removeTimes(k) >= massPlotXLim(1) && removeTimes(k) <= massPlotXLim(2)
        xline(removeTimes(k), '--m', sprintf('remove %d', removeLabels(k)), ...
            'LabelOrientation', 'horizontal', 'LabelVerticalAlignment', 'bottom');
    end
end

grid on;
xlabel('time [s]');
ylabel('estimated payload position [m]');
title('Payload position inferred from robot COM offset');
xlim(massPlotXLim);
legend('Location', 'best');

%% Save figures for easy sharing.
saveas(findobj('Type', 'figure', 'Name', 'Payload mass comparison'), ...
    fullfile(scriptDir, 'payload_mass_comparison.png'));
saveas(findobj('Type', 'figure', 'Name', 'Estimated COM coordinates'), ...
    fullfile(scriptDir, 'payload_com_coordinates.png'));
saveas(findobj('Type', 'figure', 'Name', 'Estimated COM xy trajectory'), ...
    fullfile(scriptDir, 'payload_com_xy_trajectory.png'));
saveas(findobj('Type', 'figure', 'Name', 'Payload position inferred from robot COM'), ...
    fullfile(scriptDir, 'payload_position_from_robot_com.png'));

fprintf('\nSaved figures:\n');
fprintf('  %s\n', fullfile(scriptDir, 'payload_mass_comparison.png'));
fprintf('  %s\n', fullfile(scriptDir, 'payload_com_coordinates.png'));
fprintf('  %s\n', fullfile(scriptDir, 'payload_com_xy_trajectory.png'));
fprintf('  %s\n', fullfile(scriptDir, 'payload_position_from_robot_com.png'));
